File: Stack/Python3/M739-daily-temperatures.py
# time:  O(n)
# space: O(n)

class Solution:
    def dailyTemperatures(self, temperatures):
        """
        :type temperatures: List[int]
        :rtype: List[int]
        """
        st = []
        res = [0] * len(temperatures)
        for i in range(len(temperatures)):
            while len(st) > 0 and temperatures[st[-1]] < temperatures[i]:
                j = st.pop()
                res[j] = i - j
                
            st.append(i)
            
        return res
		
# faster solution, not using stack
class Solution:
    def dailyTemperatures(self, temperatures):
        """
        :type temperatures: List[int]
        :rtype: List[int]
        """
		n = len(temperatures)
		res = [0] * n
		for i in range(n-2, -1, -1):
			j = i + 1
			while True:
				if temperatures[j] > temperatures[i]:
					res[i] = j - 1
					break
				elif res[i] == 0:
					break
				j += res[j]
				
		return res

File: Stack/C/M739-daily-temperatures.c
// time:  O(n)
// space: O(n)

/**
 * Return an array of size *returnSize.
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* dailyTemperatures(int* temperatures, int temperaturesSize, int* returnSize) {
    int* res = calloc(temperaturesSize, sizeof(int));
    *returnSize = temperaturesSize;
    int* st = calloc(temperaturesSize, sizeof(int));
    int ind = 0;
    int j = 0;
    for(int i = 0; i < temperaturesSize; i++){
        while(ind > 0 && temperatures[st[ind-1]] < temperatures[i]){
            j = st[--ind];
            res[j] = i - j;
        }
        st[ind] = i;
        ind++;
    }
    return res;
}

// faster solution

int* dailyTemperatures(int* temperatures, int temperaturesSize, int* returnSize) {
    int *stack, top, i, *res;
    
    res = (int *)malloc(sizeof(int)*temperaturesSize);
    memset(res, 0, sizeof(int)*temperaturesSize);
    
    stack = (int *)malloc(sizeof(int)*temperaturesSize);
    top = -1;
    *returnSize = temperaturesSize;
    
    for (i=1; i<temperaturesSize; ++i) {
        if (temperatures[i] <= temperatures[i-1]) {
            stack[++top] = i-1;
        } else {
            res[i-1] = 1;
            while (top >= 0 && temperatures[stack[top]] < temperatures[i]) {
                res[stack[top]] = i - stack[top];
                --top;
            }
        }
    }
    
    return res;
}